### ISSUE TITLE
docker_service doesn't pass environment to docker-compose

### ISSUE THREAD

**Issue by jctanner** [ansible/ansible#22946]
_From @davojan on October 18, 2016 11:22_

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

docker_service
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  config file = /usr/local/etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

N/A
##### SUMMARY

`docker_service` task doesn't pass environment variables defined using `environment` option at task- or playbook-level to the `docker-compose.yml` when using `project_src`.
##### STEPS TO REPRODUCE

Playbook:

``` yaml

---
- name: demo
  tasks:
  - docker_service: project_src=/tmp/compose-dir
    environment:
      MYSQL_DB: "test"
```

`docker-compose.yml` at `/tmp/compose-dir`:

``` yaml
version: '2'
services:
  mysql:
    image: mysql:5.7
  environment:
    MYSQL_DATABASE: ${MYSQL_DB}
```

##### EXPECTED RESULTS

`mysql` container should see `MYSQL_DATABASE=test` and create database "test".
##### ACTUAL RESULTS

"test" database doesn't created

But when calling docker-compose directly:

```
MYSQL_DB=test docker-compose up
```

the "test" database is created

_Copied from original issue: ansible/ansible-modules-core#5295_

**Comment by jctanner** [comment]
_From @ansibot on October 18, 2016 11:29_



@ansible ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jctanner** [comment]
_From @ansibot on November 3, 2016 8:32_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jctanner** [comment]
_From @ansibot on November 18, 2016 9:31_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jctanner** [comment]
_From @ansibot on December 7, 2016 19:49_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by jctanner** [comment]
Related:
https://github.com/ansible/ansible-modules-core/pull/3787
https://github.com/ansible/ansible-modules-core/issues/3768
https://github.com/ansible/ansible-modules-core/issues/5661

**Comment by snowsky** [comment]
Run into the same issue.

**Comment by sivel** [comment]
I think the problem is that `docker_service` would need it's own `env` argument just like `docker_container` to support this.

The `environment` directive on a task, gives the `docker_service` module access to specified environment variables, and isn't intended to give services created by `docker_service` access to those env vars.

**Comment by snowsky** [comment]
It failed the docker compose's function validate_against_config_schema(config_file) in process_config_file.

I see that the variable is expended correct:
```
{'environment': {'MYSQL_DATABASE': 'test'}, 'mysql': {'image': 'mysql:5.7'}}
```

**Comment by snowsky** [comment]
seems the compose yml file is in a wrong format:
```
version: '2'
services:
  mysql:
    image: mysql:5.7
  environment:
    MYSQL_DATABASE: ${MYSQL_DB}
```
```
Unsupported config option for services.environment
```

**Comment by snowsky** [comment]
This works for me though, looks like an indentation issue.
```
version: '2'
services:
  mysql:
    image: mysql:5.7
    environment:
      MYSQL_DATABASE: ${MYSQL_DB}
```

**Comment by ansibot** [comment]
cc @chouseknecht
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by davojan** [comment]
I've just tried ``docker_service`` again. This issue seems to be fixed and can be closed. Variables from ``environment`` (at least task-level) are perfectly available in docker-compose. Thanks!

**Comment by drzraf** [comment]
This needs to be reoponed:

$ `ansible --version`
ansible 2.5.2

$ cat test.yml 
```yaml
- hosts: localhost
  tasks:
    - name: testing compose env
      docker_service:
        debug: True
        project_name: foo
        recreate: always
        definition: 
          version: '2'
          services: { test: { image: alpine, command: "sh -vxc env", environment: { DBCOMPOSE: "FOO" } } }
      environment:
        DBTASKS: BAR
```

$ `ansible-playbook test.yml && docker ps -an1 && docker logs $(docker ps -qan1)`

## Actual:
```
changed = True
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                  PORTS               NAMES
342257782eae        alpine              "sh -c env"         1 second ago        Up Less than a second                       foo_test_1
+ env
DBCOMPOSE=FOO
HOSTNAME=c4c3c9d1ead8
SHLVL=1
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
```

## Expected:
One additional line for:
```
DBTASKS=BAR
```


This is particularly annoying when using a docker-compose file (project_src & co).

**Comment by jctanner** [comment]
@drzraf since this was a migrated issue and I was not the original submitter, it would be better for you to open a new issue that you can help carry out to resolution.

**Comment by Clement-TS** [comment]
@drzraf I reproduce the same thing with your test.

But I don't think this is an issue: unless you don't explicitly define which env var you need in your docker-compose, you won't get it in your container env.

It's the same behavior when you run `docker-compose` on your host, you don't want your build envs (`$HOME`, `$PATH`...) to be communicated to the run envs of your container.

A working example:

```test.yml 
- hosts: localhost
  tasks:
    - name: testing compose env
      docker_service:
        debug: True
        project_name: foo
        recreate: always
        definition: 
          version: '2'
          services: { test: { image: alpine, command: "sh -vxc env", environment: { DBTASKS: "${DBTASKS}" } } }
      environment:
        DBTASKS: BAR
```

```bash
$ ansible-playbook test.yml && docker ps -an1 && docker logs $(docker ps -qan1)
...
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                    PORTS               NAMES
8cb9dc08b245        alpine              "sh -vxc env"       3 seconds ago       Exited (0) 1 second ago                       foo_test_1
+ env
HOSTNAME=8cb9dc08b245
SHLVL=1
HOME=/root
DBTASKS=BAR
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
```

**Comment by davojan** [comment]
@drzraf this issue is about passing env inside docker-compose definition, NOT inside the running containers. If you want them inside the containers you should explicitly pass them in ``docker-compose.yml``.

This issue shouldn't be reopened, everything works as expected in the current version.

**Comment by drzraf** [comment]
ok
